elide-graphql/src/main/java/com/yahoo/elide/graphql/parser/GraphQLEntityProjectionMaker.java

```diff
@@ -0,0 +1,537 @@
+     */
+    public GraphQLProjectionInfo make(String query) {
+        Parser parser = new Parser();
+        Document parsedDocument = parser.parseDocument(query);
+
+        // resolve fragment definitions
+        fragmentResolver.addFragments(parsedDocument);
+
+        // resolve operation definitions
+        parsedDocument.getDefinitions().forEach(definition -> {
+            if (definition instanceof OperationDefinition) {
+                // Operations would be converted into EntityProjection tree
+                OperationDefinition operationDefinition = (OperationDefinition) definition;
+                if (operationDefinition.getOperation() == OperationDefinition.Operation.SUBSCRIPTION) {
+                    // TODO: support SUBSCRIPTION
+                    return;
+                }
+
+                // resolve variable definitions in this operation
+                variableResolver.newScope(operationDefinition);
+
+                addRootProjection(operationDefinition.getSelectionSet());
+            } else if (!(definition instanceof FragmentDefinition)) {
+                throw new InvalidEntityBodyException(
+                        String.format("Unsupported definition type {%s}.", definition.getClass()));
+            }
+        });
+
+        return new GraphQLProjectionInfo(rootLocation, rootProjections, relationshipMap);
+    }
+
+    /**
+     * Root projection would be an operation applied on an entity class. There should be only one selection, which
+     * is the entity, in the selection set. The EntityProjection tree would be constructed recursively to add all
+     * child projections.
+     *
+     * @param selectionSet a root-level selection set
+     */
+    private void addRootProjection(SelectionSet selectionSet) {
+        List<Selection> selections = selectionSet.getSelections();
+
+        Selection rootSelection = selections.get(0);
+        if (!(rootSelection instanceof Field)) {
+            throw new InvalidEntityBodyException("Entity selection must be a graphQL field.");
+        }
+
+        String entityName = ((Field) rootSelection).getName();
+        if (SCHEMA.equals(entityName) || TYPE.equals(entityName)) {
+            // '__schema' and '__type' would not be handled by entity projection
+            return;
+        }
+
+        Class<?> entityType = entityDictionary.getEntityClass(entityName);
+        if (entityType == null) {
+            throw new InvalidEntityBodyException(String.format("Unknown entity {%s}.", entityName));
+        }
+
+        rootLocation = rootSelection.getSourceLocation();
+        Field fields = (Field) rootSelection;
+
+        // merging partial graphql selections
+        selections.stream().skip(1).forEach(selection -> {
+            if (!(selection instanceof Field)) {
+                throw new InvalidEntityBodyException("Entity selection must be a graphQL field.");
+            }
+            if (!entityName.equals(((Field) selection).getName())) {
```

Comment:
I don't understand what this whole check is about.

Reply:
This is for something querying two different classes at the same time like
```
query {
  book {
    ...
  }
  country {
    ...
  }
}
```

Reply:
Let's add this example to the comment here to show why we are throwing the exception.